Added interactive navigation to extended tap results

judovana opened the pull request:
- possibility to hide/show
   - passed/failed/ignored tests
   - diagnosticd details
   - individual diagnostic items
- number is now direct url to itself

tested in firefox and chromium
The todo and bail out items ar enot covered, as I do nto have exmaple tap file with them

<!-- Please describe your pull request here. -->

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md  in your own repository 
-->

kinow commented on src/main/resources/org/tap4j/plugin/TapResult/index.jelly:
```diff
@@ -23,7 +23,14 @@
+								${it.toDo} ToDo,
+								${it.bailOuts} Bail Out!;
+								<u onclick='showHideInline("js+-")'> +/- </u>,
+							</td>
```
tabs vs. spaces

judovana replied:
The file already mixes spaces and tabs. I did some minor corrections though. Thanx!

kinow replied:
My bad :slightly_smiling_face: thanks @judovana !

kinow reviewed (commented):
Had a very quick look, and found nothing major! I'll run some tests when I find some spare time to work on the plug-in. Or if someone can test and confirm it works for them too, maybe include some screenshots, etc. That would be helpful. :slightly_smiling_face: 

Thank you for the PR!
-Bruno

kinow commented:
CI failed but it was during the git clone stage. Kicked Jenkins, waiting for it to complete, but I guess the tests should run fine.

judovana commented:
> Had a very quick look, and found nothing major! I'll run some tests when I find some spare time to work on the plug-in. Or if someone can test and confirm it works for them too, maybe include some screenshots, etc. That would be helpful. slightly_smiling_face
> 
> Thank you for the PR! -Bruno

Hmm, I may elaborate. Still the real action is the best :)  Thanx for review!

judovana commented:
Yes, tests are passing.  I hed messed up a bit with those whitespaces. I will double check tomorrow morning .it is late now here. Thaxn agai for quick review!

judovana commented:
Hello! one more set of show/hide buttons, but otherwise I'm done. If yo do nto have cycles to try it out, will attach screenshots soon.

I'm afrtaid the plugin will need some dependecy bumps:( I tried to bump it to be able to build with all jdk8,11 and 17 but failed. Simple jenkins master bump of version did not helped, and when I started to bump also other deps, I ended in testfailures:(

judovana commented:
Fresh view:
![01_fresh](https://user-images.githubusercontent.com/2904395/207549894-52fbb3b6-195c-40f9-8039-1bc3ae54533c.jpg)
the +/- "buttons" can be hidden, in case you need to copy results or similarly, by clicking to first +/-
![02_clickedPlusMinus](https://user-images.githubusercontent.com/2904395/207549903-a8e3dba0-bad3-4d06-878a-b51dae0594d3.jpg)
when clicked on  _ok_ all passes disappear (another click would make them reappear)
![03_clickedOk](https://user-images.githubusercontent.com/2904395/207549907-d048053b-ceae-474a-8520-c155a3b35762.jpg)
when clicked on  _ok skipped_ all skipps disappear (another click would make them reappear)
![04_clickedSkipped](https://user-images.githubusercontent.com/2904395/207549909-0fa110c5-febd-4024-a9d2-f09a99d03bcf.jpg)
the details can be collapsed by +/- button
![05_clickedPlusMiusNextToJdk_security3_0](https://user-images.githubusercontent.com/2904395/207549912-5dfb13c9-b67d-416a-99ee-112666ea66f9.jpg)
even individual lines of  details may be collapsed
![06_clickedPlusMiusNextToOutput](https://user-images.githubusercontent.com/2904395/207549915-6c2af74f-9611-4a66-9634-f7b8732206a3.jpg)
if you click to number, the url of page will adapt to the link directly to the test
![07_clickedNumber43](https://user-images.githubusercontent.com/2904395/207549918-b3fad2c3-bee4-47c9-87fa-d8045dbf7eb6.jpg)
if you open the link in new window, it will show directly the test with all details
![08_openedUrlFromClickOn43](https://user-images.githubusercontent.com/2904395/207549921-f48d6483-932c-40fc-8497-107a483b6946.jpg)
the "buttons" on right are cumulating several clicks in one -  failed: show just list of failures
![09_clickedFailed](https://user-images.githubusercontent.com/2904395/207549928-4842aa5f-780c-4154-85a5-191901d834b1.jpg)
  failed details: show just list of failures with details
![10_clickedFailedDetails](https://user-images.githubusercontent.com/2904395/207549934-4bda7d15-8eee-41e7-8402-0163d235785e.jpg)
  failed no details:  will collpase individual details
![11_clickedNoDetails](https://user-images.githubusercontent.com/2904395/207549936-0f496f2f-c248-47b0-8b21-6a35065a2bd9.jpg)

judovana commented:
Hello!

Is there any future for this PR?

kinow commented:
Hi. Sorry end of the year means lots of deliveries at $work (not using the plugin at $work now, so merely a volunteer). Will review and merge when I find some spare time, sorry.

judovana commented:
Sure thing. You had just been supper quick with first reply. So I got confused. No worries, and good luck whatever you are doing!

kinow reviewed (commented):
Sorry, had a bit of time to test it today, but `mvn clean install` failed for me and I didn't have enough time to troubleshoot it. Searching I didn't find a good link to explain how to fix it, so I will have to postpone reviewing this one.

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] java.nio.file.NoSuchFileException: /home/kinow/Development/java/jenkins/tap-plugin/target/classes/META-INF/exposed.stapler-beans
  	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
  	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
  	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
  	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)

```

judovana commented:
unless I committed and pushed some pom changes which do not show, this is issue with to new jdk. See:
```
 JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk mvn clean install -DskipTests
[INFO] Scanning for projects...
[WARNING] The artifact xml-apis:xml-apis:jar:2.0.2 has been relocated to xml-apis:xml-apis:jar:1.0.b2
[WARNING] The POM for org.jenkins-ci.tools:maven-hpi-plugin:jar:2.1 is missing, no dependency information available
[WARNING] Failed to build parent project for org.tap4j:tap:hpi:2.4-SNAPSHOT
[INFO] 
[INFO] ---------------------------< org.tap4j:tap >----------------------------
[INFO] Building Jenkins TAP Plugin 2.4-SNAPSHOT
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  9.966 s
```
```
JAVA_HOME=/usr/lib/jvm/java-11-openjdk mvn clean install -DskipTests
[INFO] Scanning for projects...
[WARNING] The artifact xml-apis:xml-apis:jar:2.0.2 has been relocated to xml-apis:xml-apis:jar:1.0.b2
[WARNING] The POM for org.jenkins-ci.tools:maven-hpi-plugin:jar:2.1 is missing, no dependency information available
[WARNING] Failed to build parent project for org.tap4j:tap:hpi:2.4-SNAPSHOT
[INFO] 
[INFO] ---------------------------< org.tap4j:tap >----------------------------
[INFO] Building Jenkins TAP Plugin 2.4-SNAPSHOT
...
[ERROR] java.nio.file.NoSuchFileException: /home/jvanek/git/tap-plugin/target/classes/META-INF/annotations/org.kohsuke.stapler.export.ExportedBean
[ERROR] java.nio.file.NoSuchFileException: /home/jvanek/git/tap-plugin/target/classes/META-INF/annotations/hudson.Extension
[INFO] 3 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.968 s
[INFO] Finished at: 2022-12-27T07:41:20+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.1:compile (default-compile) on project tap: Compilation failure: Compilation failure: 
[ERROR] java.nio.file.NoSuchFileException: /home/jvanek/git/tap-plugin/target/classes/META-INF/exposed.stapler-beans
[ERROR]   	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
```
```
JAVA_HOME=/usr/lib/jvm/java-17-openjdk mvn clean install -DskipTests
[INFO] Scanning for projects...
[WARNING] The artifact xml-apis:xml-apis:jar:2.0.2 has been relocated to xml-apis:xml-apis:jar:1.0.b2
[WARNING] The POM for org.jenkins-ci.tools:maven-hpi-plugin:jar:2.1 is missing, no dependency information available
[WARNING] Failed to build parent project for org.tap4j:tap:hpi:2.4-SNAPSHOT
[INFO] 
[INFO] ---------------------------< org.tap4j:tap >----------------------------
[INFO] Building Jenkins TAP Plugin 2.4-SNAPSHOT
...
ERROR] java.nio.file.NoSuchFileException: /home/jvanek/git/tap-plugin/target/classes/META-INF/annotations/org.kohsuke.stapler.export.ExportedBean
[ERROR] java.nio.file.NoSuchFileException: /home/jvanek/git/tap-plugin/target/classes/META-INF/annotations/hudson.Extension
[INFO] 3 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.628 s
[INFO] Finished at: 2022-12-27T07:42:30+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.1:compile (default-compile) on project tap: Compilation failure: Compilation failure: 
[ERROR] java.nio.file.NoSuchFileException: /home/jvanek/git/tap-plugin/target/classes/META-INF/exposed.stapler-beans
[ERROR]   	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
```
```
JAVA_HOME=/usr/lib/jvm/java-19-openjdk mvn clean install -DskipTests
[INFO] Scanning for projects...
[WARNING] The artifact xml-apis:xml-apis:jar:2.0.2 has been relocated to xml-apis:xml-apis:jar:1.0.b2
[WARNING] The POM for org.jenkins-ci.tools:maven-hpi-plugin:jar:2.1 is missing, no dependency information available
[WARNING] Failed to build parent project for org.tap4j:tap:hpi:2.4-SNAPSHOT
[INFO] 
[INFO] ---------------------------< org.tap4j:tap >----------------------------
[INFO] Building Jenkins TAP Plugin 2.4-SNAPSHOT
...
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.147 s
[INFO] Finished at: 2022-12-27T07:43:14+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.4:javadoc (default) on project tap: Execution default of goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.4:javadoc failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-javadoc-plugin:2.10.4:javadoc: java.lang.ExceptionInInitializerError: null
[ERROR] -----------------------------------------------------
[ERROR] realm =    plugin>org.apache.maven.plugins:maven-javadoc-plugin:2.10.4
[ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
```


The issue we observe on  jdk11 and 17 is valid, and is caused by misaligned depndencies in pom.
I was solving it in all my plugins about year ago, and was quite a pain to hit the correct combination to make it work in jdk11 and up. Eg:
https://github.com/judovana/jenkins-report-jck/commit/b4779c6e7f2451c2b2b9160f235b87548d50fa0a
https://github.com/judovana/jenkins-report-generic-chart-column/commit/70a7573ac0d887214eb538a97d40d402b761d248
https://github.com/judovana/text-finders-plugin/commit/b3fa8fddc3b4a958e2b0d11d88d946e9ab8f8dc1

I can elaborate a bit in this direction, but it should be different  PR.

kinow commented:
>The issue we observe on jdk11 and 17 is valid, and is caused by misaligned depndencies in pom.

Thanks for investigating!

> I was solving it in all my plugins about year ago, and was quite a pain to hit the correct combination to make it work in jdk11 and up. Eg:

Yeah, when I was working with Jenkins it was a lot easier to keep track of upcoming breaking changes. But now I work on plug-ins like this one on and off, sometimes with intervals of a year. A lot happens in that period, and normally this "lot" includes new checkstyle or spotbugs changes, new maven enforcer plugin rules, replace HTML tables by divs, move things in the UI, and other things added to the plug-in inherited build configuration that break the build :disappointed_relieved: 

Your PR looks to be in order, so I apologize as this means it will take a bit longer to be reviewed and merged, at least until I sort out my local build, @judovana .

Cheers
Bruno

judovana commented:
Thanx a lot. No problems.

I will try to look into the dependency hell. I think I had already seen both 11 and 17 exception(s). It just needs good luck with version numbers :)

judovana commented:
Maybe we can merge it and release now? Pretty Please?

kinow commented:
>Maybe we can merge it and release now? Pretty Please?

The build is still not working for me and I haven't had time to try to fix it, sorry.

judovana commented:
It should be enough if you built by jdk8. Can you?

kinow commented:
> It should be enough if you built by jdk8. Can you?

Tested with jdk8 and the build is passing now. Once I find spare time (st. eulalia parties ongoing this weekend over here, then st. valentin and carnival… plus $work/life, sorry) will try to review it and if nothing's wrong, merge & release :tada:

judovana commented:
Sure thing :) Happy hacking!

kinow commented on src/main/webapp/interactive.js:
```diff
@@ -0,0 +1,97 @@
+  set("tr_details_not ok","table-row")
+  set("tr_details_ok_SKIP","table-row")
+  set("detail_body","block")
+}
```
Missing newline.

kinow commented on src/main/webapp/interactive.js:
```diff
@@ -0,0 +1,97 @@
+function showHideBlock(clazz) {
```
Would be nice to include some comments, even though most of the code is intuitive once one reads it (or the issue description & screenshots). But for future reference might be helpful.

kinow commented on src/main/webapp/interactive.js:
```diff
@@ -0,0 +1,97 @@
+
+function showFailedDetails() {
+  set("test_ok","none")
+  set("test_not ok","table-row")
+  set("test_ok_SKIP","none")
+  set("tr_details_ok","none")
+  set("tr_details_not ok","table-row")
+  set("tr_details_ok_SKIP","none")
+  set("detail_body","block")
+}
+
+function hideDetails() {
+  set("test_ok","none")
+  set("test_not ok","table-row")
+  set("test_ok_SKIP","none")
+  set("tr_details_ok","none")
+  set("tr_details_not ok","table-row")
+  set("tr_details_ok_SKIP","table-none")
+  set("detail_body","none")
+}
+
+function showNothing() {
+  set("test_ok","none")
+  set("test_not ok","none")
+  set("test_ok_SKIP","none")
+  set("tr_details_ok","none")
+  set("tr_details_not ok","none")
+  set("tr_details_ok_SKIP","none")
+  set("detail_body","none")
+}
+
+function showAll(){
+  set("test_ok","table-row")
+  set("test_not ok","table-row")
+  set("test_ok_SKIP","table-row")
+  set("tr_details_ok","table-row")
+  set("tr_details_not ok","table-row")
+  set("tr_details_ok_SKIP","table-row")
+  set("detail_body","block")
+}
```
We could prevent [global pollution](https://www.w3.org/wiki/JavaScript_best_practices#Avoid_globals) (especially for that `set` function) with a JS module/namespace, e.g.

```suggestion
/**
 * TAP functions for interactive display.
 */
const TAP = function () {
  /**
   * Set the style of an element found by its class.
   * @private
   * @param clazz {String} - The HTML element class name to locate the element.
   * @param type {String} - The HTML CSS display of the element to be set.
   */
  function set(clazz, type) {
    const all = document.getElementsByClassName(clazz);
    for (let i = 0; i < all.length; i ++) {
      all[i].style.display = type;
    }
  }

  /**
   * Iterate each element found by the HTML CSS class name checking its
   * CSS display property. If the display property is not set, then it will
   * set it to the value of `type`. Otherwise, the display will be set
   * to `none` (i.e. it toggles `display={none|type}`).
   *
   * @private
   * @param clazz
   * @param type
   */
  function showHide(clazz, type) {
    const all = document.getElementsByClassName(clazz);
    for (let i = 0; i < all.length; i ++) {
      if (all[i].style.display==='none') {
        all[i].style.display = type
      } else {
        all[i].style.display = 'none'
      }
    }
  }

  /**
   * Toggle the display of elements with the given class using the
   * display type `block`.
   * @param clazz {String} - The CSS class name.
   */
  function showHideBlock(clazz) {
    showHide(clazz, 'block')
  }

  function showHideRow(clazz) {
    showHide(clazz, 'table-row')
  }

  function showHideRowGroup(clazz) {
    showHideRow("test_" + clazz)
    showHideRow("tr_details_" + clazz)
  }

  function showHideInline(clazz, type) {
    showHide(clazz, 'inline')
  }

  function showHideRowOne(id) {
    showHideOne(id, 'table-row')
  }

  function showHideInlineOne(id) {
    showHideOne(id, 'inline')
  }

  function showHideBlockOne(id) {
    showHideOne(id, 'block')
  }

  function showHideOne(id, type) {
    var el = document.getElementById(id);
    if (el.style.display === 'none') {
      el.style.display = type
    } else {
      el.style.display = 'none'
    }
    ;
  }

  function showFailed() {
    set("test_ok","none")
    set("test_not ok","table-row")
    set("test_ok_SKIP","none")
    set("tr_details_ok","none")
    set("tr_details_not ok","none")
    set("tr_details_ok_SKIP","none")
    set("detail_body","none")
  }

  function showFailedDetails() {
    set("test_ok","none")
    set("test_not ok","table-row")
    set("test_ok_SKIP","none")
    set("tr_details_ok","none")
    set("tr_details_not ok","table-row")
    set("tr_details_ok_SKIP","none")
    set("detail_body","block")
  }

  function hideDetails() {
    set("test_ok","none")
    set("test_not ok","table-row")
    set("test_ok_SKIP","none")
    set("tr_details_ok","none")
    set("tr_details_not ok","table-row")
    set("tr_details_ok_SKIP","table-none")
    set("detail_body","none")
  }

  function showNothing() {
    set("test_ok","none")
    set("test_not ok","none")
    set("test_ok_SKIP","none")
    set("tr_details_ok","none")
    set("tr_details_not ok","none")
    set("tr_details_ok_SKIP","none")
    set("detail_body","none")
  }

  function showAll(){
    set("test_ok","table-row")
    set("test_not ok","table-row")
    set("test_ok_SKIP","table-row")
    set("tr_details_ok","table-row")
    set("tr_details_not ok","table-row")
    set("tr_details_ok_SKIP","table-row")
    set("detail_body","block")
  }

  return {
    showHideBlock,
    showHideRow,
    showHideRowGroup,
    showHideInline,
    showHide,
    showHideRowOne,
    showHideInlineOne,
    showHideBlockOne,
    showHideOne,
  }
}()

```

kinow replied:
Didn't finish the example above.. but the idea was to "export" the public functions in the `return {]` dict/obj, and hide internal/private ones like `showHide` and `set`, also avoiding possibly conflicts with other JS globals with same names. Also added only a couple of comments to illustrate the idea.

kinow reviewed (changes requested):
Hi @judovana ! Got down with some :bug: so had to stay home the whole weekend. Good opportunity to catch up with Open Source issues.

It had been ages since I last looked at this code, and seeing some headers with the date of 2010 brings back a lot of memories (but also reminds me how better this code could be, sorry! :bow: ).

I left a few comments in the code, and had a go at testing the change. The functionality suggested works, but I think we could improve the user experience a little. See screenshots & comments below.

![image](https://user-images.githubusercontent.com/304786/219855412-7549c58d-3b99-4f2e-afe0-cc98fb5b21a9.png)

In the screenshot above, it is not immediately clear that the `1 ok`, or `failed` can be clicked. At least we would have to add a `cursor: pointer` CSS style, I think.

TODO and Bail Out are ignored and cannot be hidden/displayed. Not sure if intentional.

![image](https://user-images.githubusercontent.com/304786/219855614-b1b2b9a9-4d5c-4524-8aa4-a23efbbf5e66.png)

The Nothing when active still displays the TODO element. I assume Bail Out elements will also be displayed.

Might be good to have a help icon explaining each option, especially the "failed details" and "no details". Took me a while to realize what they were supposed to do (as my example was rather simple).

There is no indication to the user of what filters are active or not. IMO instead of clickable text, it would be better to have a toggle element, a check box, etc.

Cheers
Bruno

judovana commented:
Hello! Thanx a lot for review!

 * bailed out/todo - They are currently indeed not implemented, becasue  I lack tapset having them properly in.
   * ----------
       *  If you provide me your testing tapset, I will add the missing functionality. 
       *  What weight they have?  Are thy more close to failed, or to passed; aka how much should user be interested to them? To me, the top level view is all run. If so, then i need to see what failed, and idealy traces only. 
       *  this will be done as separate commit. it can be part of this PR, but do not need to be.

 * ternary/if+else and replace x String.format
    * ----------
         *  Ternary is awesome thing, but I would say lets use it as sparse as possibly.
         *  Similarly format x repalce are usually metter fo taste, where I incline to replace, considering it a bit more readable out of the glance
         * Unless you insists, I would like to keep this part of the code as it is.

 * thanx a lot for pointing out  [global pollution](https://www.w3.org/wiki/JavaScript_best_practices#Avoid_globals) . I definitely agree on fixing it, and will fix it. I agree this is blocker for this PR and I'm sorry Iincluded it. 
     * ----------
       *   Thanx a lot for reminder and example! It save me some dust swaping out of unused (read JS:) part of my brain

judovana commented:
Hello! Time for som eopensource youa re saying? Here it is - major showstopper of global polution shoudl be fixed.

I'm ok with merging, and  adding bailedouttodo later once you provide the tap file.
If I have overlooked anything,  please point again.

How do you see it?

Double thanx for review!

kinow commented:
>How do you see it?

Busy with projects at $work, but this PR is on my GitHub notifications, sorry for the delay! :+1:

judovana commented:
@kinow  happy to see you survived the "some 🐛" , was startig to be afraid:)
As promissed earlier, I looked a bit into the jdks compatibility  - https://github.com/jenkinsci/tap-plugin/pull/30  but therare really ancient lines in tap set plugin!

judovana commented:
"an while they discussed,  my beard grow long and grey, then I found they have not discussed,  they played pokeemons all that long time"
              --judovana

kinow reviewed (commented):
Managed to clean my GitHub notifications from $work and other pending issues, and it's Easter, so I have some spare time to look at the last issue on my GH notifications :slightly_smiling_face: 

![image](https://user-images.githubusercontent.com/304786/230778234-193c24b8-cf4d-41d2-9460-7aa264888d96.png)

Not looking at the code today, just the functionality. @judovana I think the previous feedback was not entirely fixed. There's a missing newline review feedback that's not resolved, and the text on the webpage still gives no indication to users that they are clickable.

See the GIF with the mouse pointer when it's hovered another clickable elements like links, and when it hovers over the new options to display/hide extended tap results.

![GIFrecord_2023-04-09_162634](https://user-images.githubusercontent.com/304786/230778541-b9f1eb15-0a93-4545-8cee-d8873a3dac61.gif)

The mouse pointer is actually the same when you are in a textarea, where you can select text.

Clicking on "Nothing" still displays entries that are TODO or Bail Out:

![image](https://user-images.githubusercontent.com/304786/230778954-bae1b4f9-443a-42ea-8ba0-640799c6b505.png)

The TAP Stream used:

```tap
1..4
ok 1 - Input file opened
not ok 2 - First line of the input valid
ok 3 - Read the rest of the file
not ok 4 - Summarized correctly # TODO Not written yet
```

It doesn't seem to be possible to toggle Bail out, nor Todos. For example:

![image](https://user-images.githubusercontent.com/304786/230778914-07769b0a-2384-44e4-ad38-0668909557a5.png)

Clicking on "failed details" and "no details" brought the same results for me. It's not clear how they differ (I guess it might be different if I have YAMLish in the TAP stream?), I think other users might find it confusing too.

That's it for now :wave: happy easter!

Bruno

judovana commented:
thanx! Will elaborate!

judovana commented:
Hello!
Had added the cursor, and added tooltips explianing the confusing behavior

And the new line ;)

Will now elaborate do n bailed out and todo tests with:
```
echo "1..7
ok 1 - Input file opened
not ok 2 - First line of the input valid
ok 3 - Read the rest of the file
not ok 4 - Summarized correctly # TODO Not written yet
Bail out!
Bail out! with reason" >  tap.tap
```
as input.

That may take some time..  

Please, try, and If I missed anything again, please let me know!

judovana commented:
Hello! During addition of handlng of bailouts, I found I foreget to handle comments. Will ad them.
However - I got stuck a bit: 
both those lines: https://github.com/jenkinsci/tap-plugin/blob/b75bd83b627e445966652b8532a7cda1dfa97541/src/main/resources/org/tap4j/plugin/tags/line.jelly#L37  and https://github.com/jenkinsci/tap-plugin/blob/b75bd83b627e445966652b8532a7cda1dfa97541/src/main/resources/org/tap4j/plugin/tags/line.jelly#L45
have both calls  `${it.getClazz(tapLine)}` and ` ${it.getId(tapLine,tapFile)}` not executed. So the `<tr>` lacks both `id` nad `class` .  Notice, that I'm using the same pattern also for https://github.com/jenkinsci/tap-plugin/blob/b75bd83b627e445966652b8532a7cda1dfa97541/src/main/resources/org/tap4j/plugin/tags/line.jelly#L23  where it is evaluated proeprly.

I have met similar things with jelly - no exception nowhere, but expression not evaluated. Maybe because of inheritance, maybe because of typo.  

Maybe you will see what I do not see? Thanx!

judovana commented:
Hello sir! So yet again, all requested changes should be there!

judovana commented:
pretty please?

judovana commented:
ping please?

judovana commented:
All is fixed. Please try. If no time to try, just merge. We are waiting for this for half  a yea rnow:( You have only some 120 half a years like this in life:(

judovana commented:
ping please?  I would like to return to https://github.com/jenkinsci/tap-plugin/pull/30 but I realy do not want to have two huge PRs in progress

judovana commented:
Pleaseeee?!?!?!?!?!?!?

judovana commented:
HI @kinow  all was resolved ages ago, I keep living in deployed for more then half of the year. 
I wouldlike to work on https://github.com/jenkinsci/tap-plugin/pull/30 as it is becoming crucial, but I would really like to finsih this one.

if you are alive, please do not leave this to get rotten.

judovana reviewed (commented):
ok

judovana commented:
And so the summer passed, and then autumn and winter.  After winter and another beautiful spring, and after it another summer.

And so the years were passing,  kids grow up, and old people died. New kids were born, and ages passed away..

At the end, even the see was dry, and mountains dust, but poor judovana was still living with his fork in production....

judovana commented:
Please?  if you proceed, I will fix https://www.jenkins.io/security/advisory/2023-09-06/#SECURITY-3190 for you.

judovana commented:
@kinow  Discussion moved. What now? Do you still miss anything, or have any concerns? I would be most happy to fix it.

kinow commented:
> @kinow Discussion moved. What now? Do you still miss anything, or have any concerns? I would be most happy to fix it.

Hi @judovana , nothing pending with you at the moment. I don't have time to review this one during $working hours, so I do most reviews over weekends now. This weekend I was reviewing a PR for Commons Imaging (opened in December last year, so you can see I have a backlog of PR's and issues to work through https://github.com/apache/commons-imaging/pull/254).

Some nights I find 1-2 hours to work on open source. Next weekend if I have finished this Imaging issue, then I will start syncing Jenkins repositories. But there are other Jenkins issues to be solved too (in Infra, some other security bugs [hidden, managed in some private repos by Jenkins infra], active choices plug-in, and tap-plugin).

Just trying to be transparent so you are aware that I have other open source issues, and that my priorities are work > private life > volunteering. So I cannot promise you to review it today, nor this weekend, sorry. But hopefully before the end of the year as I don't have any other big commitments nor large-period travels planned.

judovana commented:
> > @kinow Discussion moved. What now? Do you still miss anything, or have any concerns? I would be most happy to fix it.
> 
> Hi @judovana , nothing pending with you at the moment. I don't have time to review this one during $working hours, so I do most reviews over weekends now. This weekend I was reviewing a PR for Commons Imaging (opened in December last year, so you can see I have a backlog of PR's and issues to work through [apache/commons-imaging#254](https://github.com/apache/commons-imaging/pull/254)).

Nice! :clap:  Good luck with that one!
> 
> Some nights I find 1-2 hours to work on open source. Next weekend if I have finished this Imaging issue, then I will start syncing Jenkins repositories. But there are other Jenkins issues to be solved too (in Infra, some other security bugs [hidden, managed in some private repos by Jenkins infra], active choices plug-in, and tap-plugin).
> 
> Just trying to be transparent so you are aware that I have other open source issues, and that my priorities are work > private life > volunteering. So I cannot promise you to review it today, nor this weekend, sorry. But hopefully before the end of the year as I don't have any other big commitments nor large-period travels planned.

Sure, thanx a lot for honesty, Highly appreciated. Are you sure you do not want to allow me as comaintainer for this plugin? I'm using it quite a lot, and it really needs several heavy reworks, otherwise it will die out. You can always eyball it and see that I'm not doing anything evil.

kinow commented:
>Sure, thanx a lot for honesty, Highly appreciated. Are you sure you do not want to allow me as comaintainer for this plugin? I'm using it quite a lot, and it really needs several heavy reworks, otherwise it will die out. You can always eyball it and see that I'm not doing anything evil.

Thanks for the offer! Not that I don't want, it's just that I would expect contributions (docs, tests, code, small, big, etc.), to show whoever becomes a co-maintainer can maintain the code and understands how the Jenkins infra works, how to release the plug-in etc. I think the issue was that your contribution got blocked here by my lack of time to volunteer. I am sure if we had started with other smaller contributions you would possibly already be a co-maintainer :slightly_smiling_face: And then I wouldn't be looking over your work, but more working together.

judovana commented:
I had created and maintain:
  https://github.com/jenkinsci/report-jtreg-plugin
  https://github.com/jenkinsci/report-generic-chart-column-plugin
  https://github.com/jenkinsci/chartjs-api-plugin
  (not yet in jenkins org, but hopefully will be) https://github.com/judovana/jenkins-scm-koji-plugin

I had designed, and inherited onc original author withdraw from java world:
  https://github.com/jenkinsci/simple-queue-plugin/pulls

I had un-orpahned, moved to 21st century and heavily extended:
  https://github.com/jenkinsci/nested-view-plugin/

All are regularly released, so you can be sure I know what I'm doing.

kinow commented:
>All are regularly released, so you can be sure I know what I'm doing.

Sorry for not being clear. It was not about not trusting your technical ability. It is about giving maintainer status to someone who has done some collaboration (i.e. probably some code merged, released) and understood the way to maintain and release Jenkins plug-ins (e.g. backwards compatibility, serialization issues, the internal security process, etc.)

judovana commented:
sure thing. TY!

kinow reviewed (changes requested):
Hi @judovana !

I used the same heredoc scripts from the previous review, with no modifications. I actually already had the job configured, which was handy to review this PR. Apologies for not sharing the code & scripts, I saw you asked for it in another comment a long time ago. Here's what I have in my job:

```bash
cat > 1.tap <<EOF
1..4
ok 1 - Input file opened
not ok 2 - First line of the input valid
ok 3 - Read the rest of the file
not ok 4 - Summarized correctly # TODO Not written yet
EOF

cat > 2.tap <<EOF
1..2
ok 1
ok 2
EOF

cat > 3.tap <<EOF
1..2999
ok 1 - here we go
Bail out! Nope.
EOF
```

Had a quick look at the code, then started playing with the options.

1. Maybe we should make one line with the summary of the results (number of files, number of tests, number of OKs, number of Not OKs, etc.), and another line with the "Filters".

2. If I click on "not ok", that hides failed tests, but keeps the Bailed out test. If I click again on that same option, it hides failed tests (it toggles the values). If I click on "show: failed", then it show only failed, and the bailed out test. This is an inconsistency

3. From the [previous comment](https://github.com/jenkinsci/tap-plugin/pull/29#pullrequestreview-1304501113),

>There is no indication to the user of what filters are active or not. IMO instead of clickable text, it would be better to have a toggle element, a check box, etc.

![image](https://github.com/jenkinsci/tap-plugin/assets/304786/4e9ba579-4205-44cc-ba81-d09076052472)

The screenshot above has hidden elements, but you are not able to tell which ones, without doing some math and applying some logic. We can probably do better by using HTML elements like radio or check boxes instead. That means we could share screenshots of the test results filtered, and whoever looked at that would be able to tell which filters we used.

4. From the pull request template, “ Link to relevant issues in GitHub or Jira”, it needs a JIRA issue.

5.  And also “Ensure you have provided tests - that demonstrates feature works or fixes the issue”, which I learned recently in Active Choices, thanks to @rahulsom 's pull requests. I can provide the basic changes in the TAP plug-in to have a similar test harness with selenium, then someone would have to write the tests to cover these changes (after writing the first test the others should be fairly easy, I think).

6. When I hover the "no details" option, I see "hide details of each row". Reading that, I thought the "Description" would be hidden.

![image](https://github.com/jenkinsci/tap-plugin/assets/304786/09e555c1-2982-49d9-a3b9-c90009217158)

I click on that option, then I see the failed test, and the bailed out test. Not sure if that was the intention of that filter.

![image](https://github.com/jenkinsci/tap-plugin/assets/304786/7fdf6dd4-20ce-40e7-9cd9-c89719719290)

7. There might be a better way to handle showing and hiding options, instead of having to have multiple `set`'s, and `show*` functions. I will have a think about it too.

8. There is an option that I don't remember from the previous reviews (maybe I didn't see it before), which hides and shows the "show" filters.

![image](https://github.com/jenkinsci/tap-plugin/assets/304786/5a74d126-b7f0-4ff1-bc3c-8ea8c9c46a38)

If I click on that, the "show failed / failed details...." options are gone. Click again, and they are back. Which looks OK. But the description of each test also includes a +/-, but clicking it nothing happens?

There is actually an error in the browser console.

```javascript
Uncaught TypeError: el is null
    showHideOne http://localhost:8080/jenkins/plugin/tap/interactive.js:84
    showHideRowOne http://localhost:8080/jenkins/plugin/tap/interactive.js:56
    onclick http://localhost:8080/jenkins/job/test-tap1/7/tapResults/:1
```

![image](https://github.com/jenkinsci/tap-plugin/assets/304786/77c5a119-41b1-4b9c-8ccf-4dda16359846)

9. If I click on Skipped, the test with the TODO directive is not toggled (I believe both TODO and SKIP are valid skipped modifiers for tests).

![image](https://github.com/jenkinsci/tap-plugin/assets/304786/80c33374-7cbf-4f32-9636-baf15759f1ee)

(I clicked on Skipped in the screenshot above; it also doesn't account for that one in the tally results.)

---

I will start updating the plug-in code to clean up some old mess, and catch up with what's new in Jenkins since the last time I looked at the code (it will mainly use active choices as reference). It will take some time, so you should be able to work on the feedback provided here.

Cheers.

Bruno

judovana commented on src/main/java/org/tap4j/plugin/TapResult.java:
```diff
@@ -274,6 +274,14 @@ public int getBailOuts() {
         return this.bailOuts;
     }
 
+    public int getComments() {
+        int count = 0;
+        for (TestSetMap testSet: testSets) {
+            count += testSet.getTestSet().getComments().size();
+        }
+        return count;
+    }
```
It does, only streams so so weirdly bad to debug if they go wild. Do you insists?

kinow replied:
Not a blocker if not implemented :+1:

judovana commented on src/main/webapp/interactive.js:
```diff
@@ -0,0 +1,261 @@
+    setTrTodos("table-row")
+    set("_comment_", "table-row")
+    set("detail_body", "block")
+  }
```
Unluckily I do not know another way. If you know, please sahre, otherwise  I will be forced to keep this aprt as it is;(

kinow replied:
Will have to think a bit, but will try to come up with a suggestion to improve.

judovana commented:
Thanx a lot for  extensive review! Will elaborate asap.

judovana commented:
Hi! I'm finally on it. I had nonvolunteerly  moved to another project, which is not using TAPs..:(  NVM.. I would still like to finish this one. Please do not do any refactorings/mass-spotbugs fixing. It is making the rebase unnecessarily  lengthy.

I will keep introducing  single small commits, and as you pledged, it will be squashed (if you still wish to do it) during merge. TY!

judovana commented:
> I used the same heredoc scripts from the previous review, with no modifications. I actually already had the job configured, which was handy to review this PR. Apologies for not sharing the code & scripts, I saw you asked for it in another comment a long time ago. Here's what I have in my job:
> ...
Thanx, it was helpful.  In meantime, I had generate dmy own artificial tap files, and also I ahd used it, and tested in production on the tapset we get ftom AQAvit tapsets. (quite huge and compelx one).


> Maybe we should make one line with the summary of the results (number of files, number of tests, number of OKs, number of Not OKs, etc.), and another line with the "Filters".

done

> If I click on "not ok", that hides failed tests, but keeps the Bailed out test. If I click again on that same option, it hides failed tests (it toggles the values). If I click on "show: failed", then it show only failed, and the bailed out test. This is an inconsistency\

explained in https://github.com/jenkinsci/tap-plugin/pull/29#issuecomment-1795662582

> There is no indication to the user of what filters are active or not. IMO instead of clickable text, it would be better to have a toggle element, a check box, etc.

Fixed. The links are now toogle buttons with on/of and mixed state. The mixed state is necessary, as you can click to  "show all" and then user hide some manually. In that moment, the button move from on/of to middle mixed state.
Also the buttons disbales themselves if there is no family of tests to be treated (eg no bailed or no skipped...)


> The screenshot above has hidden elements, but you are not able to tell which ones, without doing some math and applying some logic. We can probably do better by using HTML elements like radio or check boxes instead. That means we could share screenshots of the test results filtered, and whoever looked at that would be able to tell which filters we used.

I had fixed like this: once you click some filter, the buttons realign themselves, so they are properly on/off (no middle state can happen here) as filter set it up



>  From the pull request template, “ Link to relevant issues in GitHub or Jira”, it needs a JIRA issue.

For features, jiras are optional. Do yuou really insists for creating "jira just for jira", when there is all info and history and conversation in this PR?

>  And also “Ensure you have provided tests - that demonstrates feature works or fixes the issue”, which I learned recently in Active Choices, thanks to @rahulsom 's pull requests. I can provide the basic changes in the TAP plug-in to have a similar test harness with selenium, then someone would have to write the tests to cover these changes (after writing the first test the others should be fairly easy, I think).

If you import the selenium suite, I will be glad to add tests. But, please, do not let it block this pr.


>    When I hover the "no details" option, I see "hide details of each row". Reading that, I thought the "Description" would be hidden.

From what I may say, your tap file do not represent the feature. In tap files - at least as tap plugin is visualizing it, are two types of messages - the "title", which is the one you see in your tap file and view. That is followed by directive. As that is part of the title row and have directive,  I had intentionally left it in.
Then there are details. There can be several details, and each of them renders special  row. Each detail have its title and its body. The body is actually span over two TD table cell elements. The "failed details" and "failed no details" operates only over those second set of details.
I will attach one of the tap files I'm using.
The old links - new buttons, now had been enhanced for two more buttons, which represent both those different messages types. You will see the difference of betheirs behavior. Note, that you may not like "details row" behavior under some circumstances, but to fix this, would quite a rework of original table structure, which I can not recommend. Besides, it is really just ocasional glitch, and easily workable by user.

> I click on that option, then I see the failed test, and the bailed out test. Not sure if that was the intention of that filter.

Correct. As I personally never met bailed out tests, I was unsure how to proceed with them. I understand them as special - critical - case of failure, thus they are shown always, unless you specifically hide "bailed out tests" (it was link before, it is button now)
if you wish, I can handle them as any other failure, but currently I do not see reasoning for that.

>    There might be a better way to handle showing and hiding options, instead of having to have multiple set's, and show* functions. I will have a think about it too.

As discussed, this is no show stopper. I'm aware about pure css show/hide https://dev-bay.com/css-show-hide-div-without-javascript/ ,  and I actually like to use it: 
https://github.com/jenkinsci/report-jtreg-plugin/blob/0a49954ddb05d75081153e7b7b9a8ed922402b8d/report-jtreg/src/main/resources/io/jenkins/plugins/report/jtreg/BuildReportExtendedPlugin/index.jelly#L62 ; but this solution have its limits, and I'm not recommending it for this usecase.
Then there is new "details" element, which I'm not familiar with, but from docs (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) it have same clumsiness as css solution..

>    There is an option that I don't remember from the previous reviews (maybe I didn't see it before), which hides and shows the "show" filters.

Yes. That is the +/- "button". It have the new shiny shape, better tooltip, and remains toggled, if used. I hope you will like it :)



> If I click on that, the "show failed / failed details...." options are gone. Click again, and they are back. Which looks OK. But the description of each test also includes a +/-, but clicking it nothing happens?
> There is actually an error in the browser console.
> 
> Uncaught TypeError: el is null
>     showHideOne http://localhost:8080/jenkins/plugin/tap/interactive.js:84
>     showHideRowOne http://localhost:8080/jenkins/plugin/tap/interactive.js:56
>     onclick http://localhost:8080/jenkins/job/test-tap1/7/tapResults/:1

Right. That was error. Fixed no. Yet again, it was specific to yours tap file, not the "big real one" I'm attaching. But I'm glad it was hit, and thus fixed.

> If I click on Skipped, the test with the TODO directive is not toggled (I believe both TODO and SKIP are valid skipped modifiers for tests).>
> (I clicked on Skipped in the screenshot above; it also doesn't account for that one in the tally results.)

Simialrly as with bailed. failed/bailed and todo/skipped have each theirs own button. Is it ok?

> Please squash the commits; I will prepare some changes, and rebasing this PR onto master will be a lot easier with a single commit.

As written a bit earlier, I will be doing individual commits, and the PR will be squeezed/squashed during merge. Ibelive it is good practice, and it is making life easier for both reviwere and contribnutor. 

> I will start updating the plug-in code to clean up some old mess, and catch up with what's new in Jenkins since the last time I looked at the code (it will mainly use active choices as reference). It will t> ake some time, so you should be able to work on the feedback provided here.

Where this is heavily necessary, please postpond it until this is in. To keep re-basing bigger changeset over moreover robotic reformatings, refactroings and re-spotbugging. 
I was not checking whether the master is now passing - newly introduced (in this project) - spotbugs, but my branch is not. If I'm the casue of new spotbugs, I will elaborate, but as far as Ihave seen, the spotbugs issues were from all over the code.
please Please PLEASE, really do not fix it while this PR is in progress. That would suck remaining spirit out of me. Once it is merged, I'm even volunteering to fix spotbugs findig afterwards.


TYVM!
  J.

The ball is in your hads now!
Attached is my "big tap"
[judovana.tap.zip](https://github.com/jenkinsci/tap-plugin/files/13299346/judovana.tap.zip)

judovana commented:
motivations screenshot: https://snipboard.io/4RTEs7.jpg

judovana commented:
hello! its december! is still end of the year valid deadline?-)

judovana commented:
ping please?

judovana commented:
@kinow  2024! Happy new year! ping please?

judovana commented:
@Kinow hello!
I can not comment below https://github.com/jenkinsci/tap-plugin/pull/36#issuecomment-1893725271 and https://github.com/jenkinsci/tap-plugin/pull/36#issuecomment-1893727335  But I have to clear up some assumptions: 

First of all please s/they/he... Its jsut me, none else, and I'm quite normal "he".

>Hi @NotMyFault I think I replied in another issue (in another org?). 

It is still the same org - jenkinsci - I had requested first time after 9 months of inactivity in project, with good intention to take over plugin I depend on where maintainer is gone. As part of the issue you were contacted, and we agreed on somehow finishing it. Yet you disappeared again.@MarkEWaite suggested to wait a two weeks, and then grant me the permissions to keep plugin viable. I had waited 3 more months before becoming unpleasant again, and bothering everybody with my tinny PR again.

Now, be sure that I do not want to maintain voluntarily any more legacy open-source projects. But **everything** is better then spend rest of my life in fork. The goal of opensource projects gathering under various foundations is, that they do not die. If maintainer leave, the organization can grant another willing soul a maintainer ship. Obviously jenkins government board is not exactly united in this.
Also note, that I fully accept the reasons why you do not want to have me as comaintainer, and while the plugin is maintained and live, I do not tend to cross the border. But once it drop to become forlorn again (and If I will need it again), I will ask again.
And be sure that I fully honour that reviewer's time is much more valuable then contributor's one. 
If you look to history of PRs.. it do not give you good credit. All ended like mine. Reviwed, and then left rotten by you so long the contributors quit (or live in theirs fork for ever)

> the answer was no based that they requested that in order to get their pull request merged. I replied to the pull request (this one, I believe, #29) with a feedback on what I expected to have so that we could merge it.

Not only that. The plugin was terribly outdated and literally unsecure. As I already said, I inherited the team who needed that - on public instance - so it was obvious it have to be acted. First things first - the js/html navigation. It was no pleasure to do, but the plugin as is, is useless on big test bases with. I deal with tesbases of 100000 tests and more. The usability of tap plugin ends somewhere around 100(see bottom)
I would be delighted to improve tap plugin enough (jenkins core, XSS...), if there would be any response from you. But with delays between rounds in length of half a year? Sorry:(

Luckily for us all, in October and  December you actually worked on that, so at least something positive come out.
However, it is terribly impolite to do bigger changes in code, which was dead for ages, while there is open bigger PR, and thus forcing contributor to constantly adapt and rebase. In many communities it can be even considered insult.
NVM. my PR is no shiny, and the rebase was not that painful. Although I literally asked to not do that, you had not even bothered to reply.

>There was a reply near x-mas break,

I had replied - fixed all you wanted - 8.11.2023 That is quite long before christmas global outage.

> but I saw that we would still need to discuss more about it as they had a different opinion on some points (which is fair, and they even included screenshots and tried to explain better): #29 (comment)

@Kinow, that is perfectly ok. If you would just drop a note "I saw that, it will need tuning", I would wait much more peacefully.

>Now, my reason for saying no there (and I am fine saying that here too), is that @judovana did not send other meaningful contributions to the plug-in, nor assisted with issues, and I am not sure if they know the plug-in API, or have other experience with Jenkins plug-ins.

Maybe I would, maybe not - but do you think there is any motivation left, when we communicate on time base like Voyger->Earth->Voyger ...Oh wait.. that is much faster :)
Still I honor your decision. It was wise and I would not do differently, but I would never leave my open project so forlorn, and I would welcome any contrbutor and do my best to make them feel welcomed. 

>Alas the tap-plugin code is quite dated, and I would prefer to add other contributors who are aware of Jenkins API and release process (which I am not following very closely to be honest) and with whom communication is easier (see the other issue where you can see that there were communication issues where there was a demand for open source volunteer work to be done asap, with the messages were borderline rude, but I believe it's due to both of us living in other cultures or speaking other languages).

If I ever said something what made you (or anybody of other from jenkins community) then I'm sorry and thanx for your acceptance of cultural  mishmash. 

> This is the second time this is requested, and I will maintain that I do not support they to be added as contributor for now. No objection if in the future that happens 👍

Thats perfectly ok, but you must accept your responsibility and become reliable maintainer again. As I wrote above - I really do not want to inherit more legacy stuff. But everything is better then fork. Every member of the community have right to contribute. If overarching organization can not fill that gap of unresponsive owner, then forks starts appearing, fork wars starts, and there is no benefit nowhere until one of the forks gain wide acceptacne over original which then dies forgotten. In meantime many time, resources, bugfixes and features are lost.

> I will keep working on bugs first, and then will add his feature when we have agreed on how it should work, in an accessible and simple way for users (and also easier to maintain), and perhaps some basic tests (which I have to check if these were added -- otherwise I can help with that).

I'm looking forward for your feedback and ideas (really, a lot I do). But please, have some mercy and do not leave my waiting in queue like the most minor citizen, while you are doing other stuff in the same codebase. Or beeing silent for months and months without saying a word.

> @judovana may file other requests to become a maintainer, but I don't think that's a good way to get their issues merged or to call attention to old issues (go through scriptler, git plugin and you will find lots of older PR's and issues -- that happens).

I have to highlight - that is not just because of the PR, but also due to the terrible outdated codebase. You have worked on that in past, so thats much better. But yet again - you are forcing me to keep using fork in production, without any reasons. The PR is fixing severe issue in usability.


> (There's also a security XSS issue that needs to be fixed, that's way higher priority than PR's for new feature

That is a *year old* CVE, and usually only https://github.com/jenkinsci/nested-view-plugin/blob/27c08752b32e6a7fa1616db05653a67bc91f462b/src/main/resources/index.jelly#L1C1 so I'm not sure why you are taking that as deffender.
If I take this PR (https://github.com/jenkinsci/tap-plugin/pull/29) by word, then it is fxing no less important thing. Only because year ago, there was quite correct communication, I was not bothering to fill the issue. Also I saw the project quite dead so rather prepared the PR with worst option to live in fork.
The issue would be "tap plugin is useless on bigger number (100+?) of tests and lack any basic in-doc linking", which is severe issue. Just try the walk through the failures in the tap file attached here. Without some html/js navigation it is terrble pain and that s the smallest tap file I found.


> -- if someone were to help with those, with updating the plug-in API, etc., and communicating well, then I would be the one pushing for that person to become a co-maintainer!)

Can you be little bit more specific (the xss is private bug). I had already offer you a help (and by PR#30 I started to work on that), but no one sane want to have several PRs to half-dead project (last commit 5 years ago before I stepped in, 4 rotten PRs...), opened.

judovana commented:
Thanx for merging the fix to the CVE! I will rebase this one to apply on your changes.

judovana commented:
Rebased in https://github.com/jenkinsci/tap-plugin/pull/38

all comments and screenshots remain valid in time of creation of https://github.com/jenkinsci/tap-plugin/pull/38
